File: Java/sorting/SelectionSort.java
package sorting;

public class SelectionSort {
    public static void main(String[] args) {
        int[] arr = SortHelper.takeInput();
        int sort = SortHelper.sort();
        selectionSort(arr, sort);
        SortHelper.printSortedArray(arr);
    }

    // Time Compelexity = O(n^2)
    public static void selectionSort(int[] arr, int sort) {
        for (int i = 0; i < arr.length - 1; i++) {
            int swapVal = i;
            for (int j = i + 1; j < arr.length; j++) {
                // For sorting in Ascending Order
                if (sort == 1) {
                    if (arr[swapVal] > arr[j])
                        swapVal = j;
                }
                // For Sorting in Descending Order
                else {
                    if (arr[swapVal] < arr[j])
                        swapVal = j;
                }
            }
            int temp = arr[swapVal];
            arr[swapVal] = arr[i];
            arr[i] = temp;
        }
    }
}

File: Rust/src/sorting/selection_sort.rs
pub use super::SortOrder;

/// Selection sort algorithm.
///
/// # Arguments
/// * `arr` - A mutable reference to an array to be sorted.
/// * `order` - Sort order.
///
/// # Examples
/// ```
/// use rust::sorting::{selection_sort, SortOrder};
///
/// let mut arr = [5, 4, 3, 2, 1];
/// selection_sort(&mut arr, SortOrder::Asc);
/// assert_eq!(arr, [1, 2, 3, 4, 5]);
/// ```
///
/// # Time complexity
/// * O(n^2)
///
/// # Space complexity
/// * O(1)
pub fn selection_sort<T: PartialOrd>(arr: &mut [T], order: SortOrder) {
    for i in 0..arr.len() - 1 {
        let mut min = i;

        for j in i + 1..arr.len() {
            match order {
                SortOrder::Asc => {
                    if arr[j] < arr[min] {
                        min = j;
                    }
                }

                SortOrder::Des => {
                    if arr[j] > arr[min] {
                        min = j;
                    }
                }
            }
        }

        arr.swap(i, min);
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_selection_sort_asc() {
        let mut arr = [5, 4, 3, 2, 1];
        selection_sort(&mut arr, SortOrder::Asc);
        assert_eq!(arr, [1, 2, 3, 4, 5]);
    }

    #[test]
    fn test_selection_sort_des() {
        let mut arr = [1, 2, 3, 4, 5];
        selection_sort(&mut arr, SortOrder::Des);
        assert_eq!(arr, [5, 4, 3, 2, 1]);
    }
}
